ihmc-java-toolkit/src/main/java/us/ihmc/tools/time/FrequencyCalculator.java

```diff
@@ -1,59 +1,78 @@
 package us.ihmc.tools.time;
 
-import us.ihmc.commons.time.Stopwatch;
-import us.ihmc.tools.UnitConversions;
+import us.ihmc.commons.thread.ThreadTools;
+import us.ihmc.log.LogTools;
 
-import java.util.ArrayDeque;
+import javax.annotation.Nullable;
+import java.util.UUID;
 
+/**
+ * A rolling frequency calculator with an optional logging thread to print the frequency once per second.
+ * The frequency is updated on each query to the frequency rather than each new event.
+ * Call {@link #ping()} on each new event.
+ * Call {@link #getFrequency()} to get the current frequency.
+ */
 public class FrequencyCalculator
 {
-   private volatile double frequency;
-   private final ArrayDeque<Double> deltas = new ArrayDeque<>();
-   private final Stopwatch stopwatch = new Stopwatch().start();
-   private final int history;
-
-   /**
-    * History of 3.
-    */
-   public FrequencyCalculator()
-   {
-      this(3);
-   }
+   private static final double NANOS_IN_A_SECOND = 1_000_000_000.0;
+
+   private long lastPingTimeNanos;
+   private int pingCount;
+   private double frequency;
+   private long elapsedNanos;
+
+   @Nullable
+   private Thread loggingThread;
+   private volatile boolean loggingThreadRunning;
 
-   /**
-    * History if one yields the latest calculation only.
-    * @param history 1 or more.
-    */
-   public FrequencyCalculator(int history)
+   public FrequencyCalculator(boolean enableLoggingThread)
    {
-      this.history = history;
+      if (enableLoggingThread)
+      {
+         String threadID = UUID.randomUUID().toString().substring(0, 5);
+
+         loggingThread = new Thread(() ->
+         {
+            loggingThreadRunning = true;
+
+            while (loggingThreadRunning)
+            {
+               LogTools.info("FrequencyCalculator[" + threadID + "] average rate: " + getFrequency());
+
+               ThreadTools.sleep(1000);
+            }
+         }, getClass().getSimpleName() + "-" + threadID);
+
+         loggingThread.start();
+      }
    }
 
-   public void ping()
+   public FrequencyCalculator()
    {
-      double elapsed = stopwatch.lap();
-      deltas.addLast(elapsed);
+      this(false);
+   }
 
-      while (deltas.size() > history)
-      {
-         deltas.removeFirst();
-      }
+   public void ping() {
+      long currentTime = System.nanoTime();
 
-      double totalElapsed = 0.0;
-      for (Double delta : deltas)
-      {
-         totalElapsed += delta;
-      }
-      frequency = UnitConversions.secondsToHertz(totalElapsed / deltas.size());
+      pingCount++;
+      elapsedNanos += currentTime - lastPingTimeNanos;
+      lastPingTimeNanos = currentTime;
```

Comment:
The previous approach would do the average over a moving window while this new version does it seems the average over the entire run, which I'd argue is not desirable. I think using a moving window is desired as it discards the initial JIT compilation slowdowns rapidly and captures if they are changes in the code being ran. When running for 10min, if the monitored code running changes for more or less computing, it'll take several minutes before being able to identify the change,

Reply:
I reset the `elapsedNanos` every 500ms, so in this way it is similar to the rolling average. Would you suggest I not do that and reintroduce an additional data structure like before?

Reply:
I changed the implementation, let me know what you think about it